From: Felipe Sateler <fsateler@debian.org>
Date: Fri, 26 Feb 2021 19:29:30 -0300
Subject: Disable autospawn by default

---
 src/pulse/client-conf.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/pulse/client-conf.c b/src/pulse/client-conf.c
index 1daaf91..bf88723 100644
--- a/src/pulse/client-conf.c
+++ b/src/pulse/client-conf.c
@@ -63,7 +63,7 @@ static const pa_client_conf default_conf = {
     .cookie_from_x11_valid = false,
     .cookie_file_from_application = NULL,
     .cookie_file_from_client_conf = NULL,
-    .autospawn = true,
+    .autospawn = false,
     .disable_shm = false,
     .disable_memfd = false,
     .shm_size = 0,
